### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by TORIFUKUKaiou**
@mnishiguchi -san

Would you please review it?
Thanks a lot for your proposal.

If a new event starts, I'm gonna add two modules.
They are;
- Event module: ex. Qiita.Events.Qiitadelika202203
- Markdown module: ex. Qiita.Events.Qiitadelika202203.Markdown

### CODE REVIEW

File: lib/qiita/events/qiita_delika_202203/markdown.ex [hunk 1]
```diff
@@ -1,36 +1,41 @@
...
-    start = DateTime.new(~D[2022-03-13], ~T[15:00:00.000], "Etc/UTC") |> elem(1)
-    ending = DateTime.new(~D[2022-04-30], ~T[15:00:00.000], "Etc/UTC") |> elem(1)
+  alias Qiita.Events.Qiitadelika202203.Markdown.Data
+  alias Qiita.Events.TableUtils
 
-    Qiita.Api.items("tag:Qiitadelika OR tag:delika")
-    |> Enum.filter(fn %{"created_at" => created_at} ->
-      Timex.between?(created_at, start, ending, inclusive: :start)
-    end)
+  def generate(items) do
```
Comment (mnishiguchi) [inline lib/qiita/events/qiita_delika_202203/markdown.ex]: This is optional but it is nice to explicitly mark a callback function with `@impl`.

```suggestion
  @impl true # or @impl Qiita.Events.Markdown
  def generate(items) do
```

Comment (mnishiguchi) [inline lib/qiita/events/qiita_delika_202203/markdown.ex]: This is optional but it is nice to explicitly mark a callback function with `@impl`.

```suggestion
  @impl true # or @impl Qiita.Events.Markdown
  def generate(items) do
```

File: lib/qiita/events/main.ex [hunk 2]
```diff
@@ -0,0 +1,9 @@
+defmodule Qiita.Events.Main do
```
Comment (mnishiguchi) [inline lib/qiita/events/main.ex]: I think `Main` can be more descriptive name. What about something like `MarkdownGenerator`? Or maybe we could move the main run function to the beheviour module. That should be fine. [GenServer](https://github.com/elixir-lang/elixir/blob/89e57882e120e77d755477b96ecfa15032dd582d/lib/elixir/lib/gen_server.ex) module defines some functions as well as behavour callbacks.

```elixir
defmodule Qiita.Events.Markdown do
  @type items :: list()

  @callback generate(items) :: String.t()
  
  import Qiita.Events.Repo

  def run(query, start_time, end_time, mod, tags, title, id) do
    fetch_items(query, start_time, end_time)
    |> mod.generate()
    |> post_markdown(tags, title, id)
  end
end
```

Comment (mnishiguchi) [inline lib/qiita/events/main.ex:1]: CleanShot 2022-03-21 at 18 23 33

**Review by mnishiguchi — APPROVED** [review]
LGTM! I left a few ideas but they are minor things.

**Comment by TORIFUKUKaiou** [comment]
@mnishiguchi -san

Thank you very much for your comment.
Would you please re-review it?

File: lib/qiita/events/markdown_generator.ex [hunk 3]
```diff
@@ -0,0 +1,17 @@
+defmodule Qiita.Events.MarkdownGenerator do
+  @type items :: list()
+
+  @callback generate(items) :: String.t()
+
+  defmacro __using__(_opts) do
+    quote do
+      import Qiita.Events.Repo
+
```
Comment (mnishiguchi) [inline lib/qiita/events/markdown_generator.ex:11]: OK. If you choose to use a macro, probably you could include more common code here, such as `@behaviour Qiita.Events.MarkdownGenerator`. Maybe `import Qiita.Events.TableUtils` as well if it is considered as generic. But these are totally optional and depend on situations.

```elixir
  defmacro __using__(_opts) do
    quote do
      @behaviour Qiita.Events.MarkdownGenerator
      import Qiita.Events.Repo
      import Qiita.Events.TableUtils
```

Other than that, all looks good to me.

P.S. I personally avoid using a macro but it is my personal thing. I guess this macro is nice and simple and it tidies up code like Phoenix.

**Review by mnishiguchi — APPROVED** [review]
👍

File: lib/qiita/events/qiita_delika_202203/markdown.ex (hunk 1 above)
Comment (TORIFUKUKaiou) [reply lib/qiita/events/qiita_delika_202203/markdown.ex]: Thanks!

lib/qiita/events/qiita_delika_202203/markdown.ex has been deleted for me.

Instead, I add `@impl true` there.
https://github.com/TORIFUKUKaiou/hello_nerves/blob/d370b9d0b8c196fbfa65a68e26a441b2464fd638/lib/qiita/events/qiita_delika_202203/markdown_generator.ex#L12-L18

File: lib/qiita/events/markdown_generator.ex (hunk 3 above)
Comment (TORIFUKUKaiou) [reply lib/qiita/events/markdown_generator.ex:11]: @mnishiguchi -san

Yes.
I avoid using a macro.

But the below code happens  an error to compile.

```elixir
defmodule Qiita.Events.MarkdownGenerator do
  @type items :: list()

  @callback generate(items) :: String.t()
  
  import Qiita.Events.Repo

  def run(query, start_time, end_time, tags, title, id) do
    fetch_items(query, start_time, end_time)
    |> generate()
    |> post_markdown(tags, title, id)
  end
end
```

I desire the above code.
But I met CompileError.

```
** (CompileError) iex:10: undefined function generate/1 (expected Qiita.Events.MarkdownGenerator to define such a function or for it to be imported, but none are available)
```

I see [GenServer](https://github.com/elixir-lang/elixir/blob/v1.13.3/lib/elixir/lib/gen_server.ex#L741-L843).
So I think it uses `defmacro`.
I maybe have a misunderstanding.
Please give me a hint to solve.

**Pull request merged**